### PR TITLE
Network direction processor

### PR DESCRIPTION
**Pull request by andrewstucki**
This PR computes the ECS perimeter-based network direction of a document given a source and destination ip and a list of ip addresses that specify the internal network boundary. It's essentially a port of the [beats processor](https://www.elastic.co/guide/en/beats/filebeat/master/add-network-direction.html) along with [named range support](https://www.elastic.co/guide/en/beats/filebeat/master/defining-processors.html#condition-network).

```zsh
➜  ~ curl -H "Content-Type: application/json" -X POST -u elastic:password http://localhost:9200/_ingest/pipeline/_simulate\?verbose --data-binary @- << EOF
{
  "pipeline": {
    "processors": [
      {
        "network_direction": {
          "internal_networks": [ "private" ]
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "source": {
          "ip": "128.232.110.120"
        },
        "destination": {
          "ip": "192.168.1.1"
        }
      }
    }
  ]
}
EOF
{"docs":[{"processor_results":[{"processor_type":"network_direction","status":"success","doc":{"_index":"_index","_id":"_id","_source":{"destination":{"ip":"192.168.1.1"},"source":{"ip":"128.232.110.120"},"network":{"direction":"inbound"}},"_ingest":{"pipeline":"_simulate_pipeline","timestamp":"2020-12-18T20:23:46.589327Z"}}}]}]}%
```

See `"network":{"direction":"inbound"}`

### CODE REVIEW

**Comment by danhermann** [comment]
@andrewstucki, thanks for submitting this. I'll review it from the ES Core Features side, but it might wait until early January since our team has limited bandwidth until then.

**Comment by andrewstucki** [comment]
@danhermann that's perfectly fine, I'm hoping it can get in by 7.12 so that it can unblock some ECS upgrade work we postponed on the beats ingest side of things, but no rush. Just whipped this up to gain a bit more familiarity with ES.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@andrewstucki, there are some checkstyle violations here that need to be addressed. You should be able to see them either in the output of the `elasticsearch-ci/2` build or by running `./gradlew :x-pack:plugin:ingest:checkstyleTest` locally.

**Comment by andrewstucki** [comment]
@danhermann sorry about the late response. I merged latest master in so this wasn't stale anymore. The checkstyle issues seemed to be outdated and related to a `.*` import that I initially had. For some reason I don't think CI picked up the change when I fixed it. Hopefully we'll see the build pass this time around.

File: x-pack/plugin/ingest/src/main/java/org/elasticsearch/xpack/ingest/NetworkDirectionProcessor.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    private boolean isLoopback(InetAddress ip) {
+        return ip.isLoopbackAddress();
+    }
+
+    private boolean isUnicast(InetAddress ip) {
+        return !Arrays.equals(ip.getAddress(), BROADCAST_IP4)
+            && !isUnspecified(ip)
+            && !isLoopback(ip)
+            && !isMulticast(ip)
+            && !isLinkLocalUnicast(ip);
```
Comment (danhermann) [inline x-pack/plugin/ingest/src/main/java/org/elasticsearch/xpack/ingest/NetworkDirectionProcessor.java:179]: Elasticsearch has a convention that we don't use `!` to negate boolean values. Instead, we use the form `booleanValue == false`.

File: x-pack/plugin/ingest/src/main/java/org/elasticsearch/xpack/ingest/NetworkDirectionProcessor.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        var address = ip.getAddress();
+        return Arrays.equals(UNDEFINED_IP4, address) || Arrays.equals(UNDEFINED_IP6, address);
+    }
+
+    private boolean isPrivate(String ip) {
+        return CIDRUtils.isInRange(ip, "10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16", "fd00::/8");
+    }
+
+    private boolean isPublic(String ip) {
+        return !isLocalOrPrivate(ip);
```
Comment (danhermann) [inline x-pack/plugin/ingest/src/main/java/org/elasticsearch/xpack/ingest/NetworkDirectionProcessor.java:208]: Same as above.

**Review by danhermann — COMMENTED** [review]
@andrewstucki, this looks good. I added two very minor comments below and once those are addressed, I can get this merged.

**Comment by danhermann** [comment]
Thanks, @andrewstucki. This looks good now and I'll get it merged into the appropriate branches.

**Comment by danhermann** [comment]
cc: @elastic/es-ui in case Kibana auto-complete needs to be updated with this new processor.

**Pull request merged**